File: selfstudy/golang/cars_max.go
/*
task 222:
Cars - Max
You need to find the maximum value among car speeds in km/h
*/

package main

import (
	"fmt"
	"reflect"
)

func HighestCars(speeds []int) int {
	if len(speeds) == 0 { return 0 }
	m := speeds[0]
	for _, v := range speeds {
		if v > m { m = v }
	}
	return m
}

func main() {
	tests := []struct {
		input []int
		expected int
	}{
		{[]int{10, 20, 30}, HighestCars([]int{10, 20, 30})},
		{[]int{50, 60, 70}, HighestCars([]int{50, 60, 70})},
		{[]int{1, 2, 3}, HighestCars([]int{1, 2, 3})},
		{[]int{100, 200}, HighestCars([]int{100, 200})},
		{[]int{}, HighestCars([]int{})},
	}
	
	for idx, t := range tests {
		res := HighestCars(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// done

File: selfstudy/js/cars_max.js
/*
task 22:
Cars - Max
You need to find the maximum value among car speeds in km/h
*/

function highest_cars(speeds) {
  return speeds.length ? Math.max(...speeds) : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: highest_cars([10, 20, 30]) },
  { input: [50, 60, 70], expected: highest_cars([50, 60, 70]) },
  { input: [1, 2, 3], expected: highest_cars([1, 2, 3]) },
  { input: [100, 200], expected: highest_cars([100, 200]) },
  { input: [], expected: highest_cars([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('highest_cars', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(highest_cars(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, highest_cars(t.input) === t.expected || JSON.stringify(highest_cars(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// done
